eeps/eep-0072.md

```diff
@@ -0,0 +1,195 @@
+them.
+
+Unique error codes can be leveraged by IDEs and language servers to
+provide better contextual information about errors and make errors
+easier to search and reference. A standardized error index creates a
+common space for the Community to provide extra examples and
+documentation, creating the perfect companion for the Erlang User
+Manual and standard documentation.
+
+Rationale
+=========
+
+The concept of an "Error Index" for a programming language is not a
+novel idea. Error catalogues already exist, for example, in the
+[Rust][] and [Haskell][] Communities.
+
+Producing meaningful error messages can sometimes be challenging for
+developer tools such as compilers and type checkers due to various
+constraints, including limited context and character count.
+
+By associating a **unique code** to each _diagnostic_ (warning or
+error) we relief tools from having to condense a lot of textual
+information into a - sometime cryptic - generic, single
+sentence. Furthermore, as specific wording of errors and warnings is
+improved over time, error codes remain constant, providing a
+search-engine friendly way to index and reference diagnostics.
+
+An good example of this is the _expression updates a literal_ error
+message, introduced in OTP 27. Given the following code:
+
+    -define(DEFAULT, #{a => 1}).
+
+    updated(Value) ->
+      ?DEFAULT#{a => Value}.
+
+The compiler emits the following error:
+
+    test.erl:8:11: Warning: expression updates a literal
+        %    8|   ?DEFAULT#{timeout => 1000}.
+        %     |           ^
+
+The meaning of the error may not be obvious to everyone. Most
+importantly, the compiler provide no information on why the warning is
+raised and what a user could do about it. The user will then have to
+recur to a search engine, a forum or equivalent to proceed.
+
+Conversely, we can associate a unique identifier to the code (say,
+`ERLC-1234`):
+
+    test.erl:8:11: Warning: expression updates a literal (ERLC-1234)
+        %    8|   ?DEFAULT#{timeout => 1000}.
+        %     |           ^
+
+The code make it possible to link the error message to an external
+resource (a wiki), which contains all the required, additional,
+information about the error that would not be practical to present
+directly to the user. Here is an example of what the entry could look
+like for the above code:
+
+![Erlang Error Index Sample Entry][]
+
+Unique error codes also have the advantage to be better searchable in
+forums and chats, where the exact error message could vary, but the
+error code would be the same.
+
+Finally, error codes can be used by IDEs (e.g. via language servers)
+to match on error codes and provide contextual help. Both the [Erlang
+LS][] and the [ELP][] language server already use "unofficial" error
+codes.
+
+Format
+------
+
+An error code should be composed by two parts: an alphanumeric
+_namespace_ (three letters) and a numeric identifier (four digits),
```

Comment:
Should we relax this requirement and allow a dynamic number of chars/digits?

Reply:
We can relax it later if needed.